query_executor.go

```diff
@@ -30,12 +38,74 @@ func (q *queryExecutor) attemptQuery(qry ExecutableQuery, conn *Conn) *Iter {
 }
 
 func (q *queryExecutor) executeQuery(qry ExecutableQuery) (*Iter, error) {
-	rt := qry.retryPolicy()
+
+	// check if the query is not marked as idempotent, if
+	// it is, we force the policy to NonSpeculative
+	sp := qry.speculativeExecutionPolicy()
+	if !qry.IsIdempotent() {
+		sp = NonSpeculativeExecution{}
+	}
+
+	results := make(chan queryResponse, 1)
+	stop := make(chan struct{})
+	defer close(stop)
+	var specWG sync.WaitGroup
+
+	// Launch the main execution
+	specWG.Add(1)
+	go q.run(qry, &specWG, results, stop)
+
+	// The speculative executions are launched _in addition_ to the main
+	// execution, on a timer. So Speculation{2} would make 3 executions running
+	// in total.
+	go func() {
+		// Handle the closing of the resources. We do it here because it's
+		// right after we finish launching executions. Otherwise clearing the
+		// wait group is complicated.
+		defer func() {
+			specWG.Wait()
+			close(results)
+		}()
+
+		// setup a ticker
+		ticker := time.NewTicker(sp.Delay())
+		defer ticker.Stop()
+
+		for i := 0; i < sp.Attempts(); i++ {
+			select {
+			case <-ticker.C:
+				// Launch the additional execution
+				specWG.Add(1)
+				go q.run(qry, &specWG, results, stop)
+			case <-qry.GetContext().Done():
+				// not starting additional executions
+				return
```

Comment:
Is it possible that return happens here and `specWG.Wait()` waits forever because the `for` hasn't done all the loops?

Reply:
No, it won't. We actually already have at least 1 execution added to the waiting group in line 55 (the main execution).

Reply:
Yep, I see